lib/virtual.js

```diff
@@ -12,28 +33,73 @@ function VirtualModulesPlugin(compiler) {
 	this.compiler = compiler;
 
 	if (!compiler.inputFileSystem._writeVirtualFile) {
-		var originalPurge = compiler.inputFileSystem.purge;
+		const originalPurge = compiler.inputFileSystem.purge;
 
 		compiler.inputFileSystem.purge = function() {
 			if (originalPurge) {
-				originalPurge.call(this, arguments);
+				originalPurge.apply(this, arguments);
 			}
 			if (this._virtualFiles) {
-				Object.keys(this._virtualFiles).forEach(
-					function(file) {
-						var data = this._virtualFiles[file];
-						setData(this._statStorage, file, [null, data.stats]);
-						setData(this._readFileStorage, file, [null, data.contents]);
-					}.bind(this)
-				);
+				Object.keys(this._virtualFiles).forEach(function(file) {
+					const data = this._virtualFiles[file];
+					this._writeVirtualFile(file, data.stats, data.contents);
+				}.bind(this));
 			}
 		};
 
 		compiler.inputFileSystem._writeVirtualFile = function(file, stats, contents) {
+			const statStorage = getStatStorage(this);
+			const fileStorage = getFileStorage(this);
+			const readDirStorage = getReadDirBackend(this);
 			this._virtualFiles = this._virtualFiles || {};
 			this._virtualFiles[file] = { stats: stats, contents: contents };
-			setData(this._statStorage, file, [null, stats]);
-			setData(this._readFileStorage, file, [null, contents]);
+			setData(statStorage, file, createWebpackData(stats));
+			setData(fileStorage, file, createWebpackData(contents));
+			const segments = file.split(/[\\/]/);
+			let count = segments.length - 1;
+			const minCount = segments[0] ? 1 : 0;
+			while (count > minCount) {
+				const dir = segments.slice(0, count).join(path.sep) || path.sep;
+				try {
+					compiler.inputFileSystem.readdirSync(dir);
+				} catch (e) {
+					const time = new Date();
+					const timeMs = time.getTime();
+					const dirStats = new VirtualStats({
+						dev: 8675309,
+						nlink: 0,
+						uid: 1000,
+						gid: 1000,
+						rdev: 0,
+						blksize: 4096,
+						ino: inode++,
+						mode: 16877,
```

Comment:
why is mode different here than line 126? can we refactor to share the same `VirtualStats` or at least `VirtualStats` options?